sdk/eventhubs/azure_messaging_eventhubs/src/common/recoverable_connection.rs

```diff
@@ -0,0 +1,1055 @@
+                cbs.attach().await?;
+                Ok(cbs)
+            },
+            &self.retry_options,
+            Some(Self::should_retry_cbs_response),
+        )
+        .await?;
+        debug!("AMQP Claims-Based Security (CBS) client ensured.");
+        Ok(cbs_client)
+    }
+
+    async fn ensure_receiver(
+        self: &Arc<Self>,
+        source_url: &Url,
+        message_source: &AmqpSource,
+        receiver_options: &AmqpReceiverOptions,
+    ) -> Result<Arc<AmqpReceiver>> {
+        let mut receiver_instances = self.receiver_instances.lock().await;
+        if !receiver_instances.contains_key(source_url) {
+            self.ensure_connection().await?;
+            self.authorizer.authorize_path(self, source_url).await?;
+
+            let session = self.get_session(source_url).await?;
+
+            debug!("Create receiver on partition {source_url}.");
+            let receiver = AmqpReceiver::new();
+            receiver
+                .attach(
+                    &session,
+                    message_source.clone(),
+                    Some(receiver_options.clone()),
+                )
+                .await?;
+
+            receiver_instances.insert(source_url.clone(), Arc::new(receiver));
+        }
+
+        Ok(receiver_instances
+            .get(source_url)
+            .ok_or_else(|| EventHubsError::from(ErrorKind::MissingMessageReceiver))?
+            .clone())
+    }
+
+    async fn ensure_sender(self: &Arc<Self>, path: &Url) -> Result<Arc<AmqpSender>> {
+        let mut sender_instances = self.sender_instances.lock().await;
+        if !sender_instances.contains_key(path) {
+            // Ensure that we are authorized to access the senders path.
+            self.authorizer.authorize_path(self, path).await?;
+
+            // Retrieve a session for the sender from the session cache.
+            let session = self.get_session(path).await?;
+            let sender = AmqpSender::new();
+            sender
+                .attach(
+                    &session,
+                    format!(
+                        "{}-rust-sender",
+                        self.application_id
+                            .as_ref()
+                            .unwrap_or(&DEFAULT_EVENTHUBS_APPLICATION.to_string())
+                    ),
+                    path.to_string(),
+                    None,
+                )
+                .await?;
+            sender_instances.insert(path.clone(), Arc::new(sender));
+        }
+
+        Ok(sender_instances
+            .get(path)
+            .ok_or_else(|| EventHubsError::from(ErrorKind::MissingMessageSender))?
+            .clone())
+    }
+
+    fn should_retry_management_response(e: &azure_core::Error) -> bool {
+        match e.kind() {
+            AzureErrorKind::Amqp => {
+                warn!("Amqp operation failed: {}", e.source().unwrap());
+                if let Some(e) = e.source() {
+                    debug!("Error: {}", e);
```

Comment:
Here and elsewhere below you're "flattening" the error to just a string, which often isn't that helpful. Sure you don't *also* want to attach it using its `Debug` implementation e.g.,

```rust
debug!(err = ?e, "Error: {e}");
```

Reply:
I'd not run into that construct before, nice.